In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse=pd.read_csv('adverse_reactions.csv')
tcut=pd.read_csv('treatments_cut.csv')

In [3]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)
print(tcut.shape)

(503, 14)
(280, 7)
(34, 3)
(70, 7)


In [4]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [5]:
treatment.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [6]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [7]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [8]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [9]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 944.0+ bytes


In [10]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [11]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [12]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [13]:
patients.sample(5)

patient_id assigned_sex given_name        surname              address  \
502         503         male        Pat        Gersten    2778 North Avenue   
414         415         male    Bartosz        Wysocki   1079 Ingram Street   
265         266       female     Þórunn  Tryggvadóttir     77 Juniper Drive   
309         310         male      Valur      Bjarkason  1782 New Creek Road   
318         319         male     Benoît         Bonami   230 Lady Bug Drive   

            city     state  zip_code        country  \
502         Burr  Nebraska   68324.0  United States   
414   Versailles        OH   45380.0  United States   
265          Mio        MI   48647.0  United States   
309   Birmingham        AL   35203.0  United States   
318  Garden City  New York   11530.0  United States   

                                            contact   birthdate  weight  \
502            PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2   
414             937-526-1152BartoszWysocki@cuvox.de    4/2/1951   148.7   
265  orunnTryggvadottir@dayrep.com+1 (989) 826-7884    6/9/1978   120.0   
309             256-872-9211ValurBjarkason@cuvox.de  12/18/1933   130.7   
318            BenoitBonami@gustr.com1 718 954 8136   4/13/1975   162.8   

     height   bmi  
502      71  19.3  
414      66  24.0  
265      69  17.7  
309      68  19.9  
318      70  23.4

In [14]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

### Quality

[Completeness, Validity, Accuracy, Consistency]

#### `Patients`

- Zip code col is stored in decimal format[V]
- Certain zip codes of 4 digits[V] 
- For the state col sometimes full names are use and other times short forms are used[IC]
- Missing values for contact details of 12 patients[C]
- Incorrect data types for assigned_sex and birthdate cols[?]
- Erroneous data for weights and heights[Accuracy]
- Inconsistent way of storing phone numbers in contact col[IC]
- Duplicate and not useful data in dataset[IC]

#### `Treatments`

- hba1c_change col has missing values[C]
- Novodra and Auralin cols has u attached to numbers[V]
- Only 280 rows are there, instead of 350[C]
- Name and surname col is in lower case[IC]
- Incorrect data type for auralin and novadra cols[V]
- hba1c_change 9 instead of 4[Accuracy]
- use of dashes instead of Nan in aurolin and novodra cols[IC]

#### `Adverse Reactions`

- Name and surname col is in lower case[IC]

### Tidiness

#### `Patients`

- Phone and email together in one col i.e. contact

#### `Treatments`

- 3 variables stored in auralin and novodra cols

#### `Adverse Reactions`

- No point of using this table


# Cleaning Data

You should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define & code 
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [15]:
# Creating copies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()

## A. Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change col has missing values

#### Define and Code

In [16]:
# Removing 12 patients whose contact details can not be found.
patients_copy.dropna(inplace=True)

In [17]:
# Appending DataFrame treatment_cut in treatment_copy to fullfill 70 missing patients.
treatment_copy=treatment_copy.append(tcut,ignore_index=True)
del(tcut)

In [18]:
# Subtract hba1c_start from hba1c_end to get hba1c_change
treatment_copy.drop(columns={'hba1c_change'},inplace=True)
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

#### Test

In [19]:
'''print(treatment_copy.shape)
print(treatment_copy.info())
print(treatment_copy.describe())'''

'print(treatment_copy.shape)\nprint(treatment_copy.info())\nprint(treatment_copy.describe())'

## B. Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Merge the `adverse` table to `treatment` table

### Define and Code

In [20]:
# Novodra and Aurolin using the melt function
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

treatment_copy=treatment_copy[treatment_copy['dose']!='-']

treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str

treatment_copy.drop(columns={'dose'}, inplace=True)

treatment_copy['start']=treatment_copy['start'].str.replace('u','')

treatment_copy['end']=treatment_copy['end'].str.replace('u','')

treatment_copy['start']=treatment_copy['start'].astype('int')

treatment_copy['end']=treatment_copy['end'].astype('int')

In [21]:
# Merging the `adverse` table to `treatment` table using pd.merge() function.
# as information of advarse reaction doesn't need a separate table.

treatment_copy=treatment_copy.merge(adverse_copy,on="given_name",how="left")

treatment_copy.drop(columns={'surname_y'},inplace=True)

treatment_copy.rename(columns={'given_name':'fname','surname_x':'lname'},inplace=True)

del(adverse_copy)

del(adverse)

In [22]:
# Extracting the phone number and email address from contact column using re (regular expression).

phone_numbers=[]
emails=[]
for ph in patients_copy['contact']:
    try:
        match=re.findall(r'\d{3}-\d{3}-\d{4}',ph)[0]
        phone_numbers.append(match)
    except:
        try:
            match=re.findall(r'\d{3} \d{3} \d{4}',ph)[0]
            match='-'.join(match.split())
            phone_numbers.append(match)
        except:
            try:
                match=re.findall(r'\(\d{3}\) \d{3}-\d{4}',ph)[0]
                match1=re.findall(r'\(\d{3}\)',match)[0][1:-1]
                match2=re.findall(r'\d{3}-\d{4}',match)[0]
                
                match=match1+'-'+match2
                phone_numbers.append(match)
            except:
                phone_numbers.append(np.nan)
    try:
        match=re.findall(r'[a-zA-z]\w{3,50}@[a-z]{2,20}.[a-z]{2,3}',ph)[0]
        emails.append(match)
    except:
        emails.append(np.nan)

In [23]:
# Inserting the extracted phone number and email into the treatment_copy dataframe and removing the original 'contact' column.

patients_copy['email']=emails

patients_copy['phone_number']=phone_numbers

patients_copy.drop(columns='contact',inplace=True)

#### Test

In [24]:
patients_copy.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi                      email  phone_number  
0      66  19.6   ZoeWellish@superrito.com  951-719-9170  
1      66  19.2       PamelaSHill@cuvox.de  217-569-3204  
2      71  24.8       JaeMDebord@gustr.com  402-363-6804  
3      70  31.7  PhanBaLiem@jourrapide.com  732-636-8246  
4      27  26.1        TimNeudorf@cuvox.de  334-515-7487

## C. Validation issues
- Zip code is in float format.
- some zip codes are less than 5 digits ( it is inconsistency as well as validity issue. )

### Code 

In [25]:
patients_copy['zip_code']=patients_copy['zip_code'].astype('int')

zipcode=[]
for i in patients_copy['zip_code']:
    x=list(str(j) for j in str(i))
    if len(x)<5:
        x.insert(0,'0')
    x=''.join(x)
    zipcode.append(x)
    
patients_copy['zip_code']=zipcode

### Test

In [26]:
patients_copy.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
4             Dothan          AL    36303  United States  2/18/1928   192.3   

   height   bmi                      email  phone_number  
0      66  19.6   ZoeWellish@superrito.com  951-719-9170  
1      66  19.2       PamelaSHill@cuvox.de  217-569-3204  
2      71  24.8       JaeMDebord@gustr.com  402-363-6804  
3      70  31.7  PhanBaLiem@jourrapide.com  732-636-8246  
4      27  26.1        TimNeudorf@cuvox.de  334-515-7487

## D. Accuracy issues
- Erroneous data for height column

### Define

Generally, Average height of an American is arround 58-69 inches. below that, it's unusual. So, patients whose height is below 58 will be removed from the table for the sake of mathematical algorithms, 

### Code

In [27]:
patients_copy[patients_copy['height']<58]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      27  26.1   

                 email  phone_number  
4  TimNeudorf@cuvox.de  334-515-7487

In [28]:
patients_copy.drop(index=4,inplace=True)

### Test

In [29]:
patients_copy[patients_copy['height']<58]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, email, phone_number]
Index: []

## E. Inconsistency Issues
- For the state col sometimes full names are use and other times short forms are used
- Duplicate and not useful data in dataset
- Name and surname col is in lower case

### Define
For the state col sometimes full names are use and other times short forms are used, So full name of states are replaced by it's short form using pd.rename() function.

### Code

In [30]:
patients_copy['state'].unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'Florida', 'NV', 'CA',
       'MO', 'New York', 'MI', 'TN', 'VA', 'AL', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       'VT', 'ID', 'DC', 'AR'], dtype=object)

In [31]:
patients_copy['state'].replace({'California':'CA','Illinois':'IL','Nebraska':'NE','Florida':'FL'},inplace=True)

### Check or Test

In [32]:
patients_copy['state'].unique()

array(['CA', 'IL', 'NE', 'NJ', 'FL', 'NV', 'MO', 'New York', 'MI', 'TN',
       'VA', 'AL', 'OK', 'GA', 'MT', 'MA', 'NY', 'NM', 'LA', 'PA', 'CO',
       'ME', 'WI', 'SD', 'MN', 'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY',
       'DE', 'MD', 'AZ', 'TX', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV',
       'RI', 'NH', 'OR', 'VT', 'ID', 'DC', 'AR'], dtype=object)

### Define
Name and surname col is in lower case in 'treatment_copy' table. So, it will be capitalized.

### CoDe

In [33]:
treatment_copy['fname']=treatment_copy['fname'].str.capitalize()

In [34]:
treatment_copy['lname']=treatment_copy['lname'].str.capitalize()

### Test

In [35]:
treatment_copy.head()

fname       lname  hba1c_start  hba1c_end  hba1c_change treatment  \
0  Veronika    Jindrová         7.63       7.20          0.43   auralin   
1      Skye  Gormanston         7.97       7.62          0.35   auralin   
2    Sophia      Haugen         7.65       7.27          0.38   auralin   
3     Eddie      Archer         7.89       7.55          0.34   auralin   
4      Asia     Woźniak         7.76       7.37          0.39   auralin   

   start  end adverse_reaction  
0     41   48              NaN  
1     33   36              NaN  
2     37   42              NaN  
3     31   38              NaN  
4     30   36              NaN

### Define
Duplicate and not useful data in dataset. So, it will be removed.

In [36]:
treatment_copy[treatment_copy.duplicated()]

fname lname  hba1c_start  hba1c_end  hba1c_change treatment  start  end  \
62  Joseph   Day          7.7       7.19          0.51   auralin     29   36   

   adverse_reaction  
62     hypoglycemia

In [37]:
patients_copy[patients_copy.duplicated()]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, email, phone_number]
Index: []

In [38]:
treatment_copy.drop_duplicates(inplace=True)

### Test

In [39]:
treatment_copy[treatment_copy.duplicated()]

Empty DataFrame
Columns: [fname, lname, hba1c_start, hba1c_end, hba1c_change, treatment, start, end, adverse_reaction]
Index: []

### Some finishing touch.

In [40]:
# reseting index number.
patients_copy.reset_index(inplace=True)
patients_copy.drop(columns={'index'},inplace=True)

In [41]:
# tweaking column names.
patients_copy.rename(columns={'assigned_sex':'sex','given_name':'fname','surname':'lname','phone_number':'phone'},inplace=True)

## Cleaned DataFrames

In [42]:
patients=patients_copy.copy()
treatment=treatment_copy.copy()

In [43]:
patients.head()

patient_id     sex   fname    lname                    address  \
0           1  female     Zoe  Wellish       576 Brown Bear Drive   
1           2  female  Pamela     Hill  2370 University Hill Road   
2           3    male     Jae   Debord      1493 Poling Farm Road   
3           4    male    Liêm     Phan        2335 Webster Street   
4           6    male  Rafael    Costa         1140 Willis Avenue   

                city state zip_code        country  birthdate  weight  height  \
0  Rancho California    CA    92390  United States  7/10/1976   121.7      66   
1          Armstrong    IL    61812  United States   4/3/1967   118.8      66   
2               York    NE    68467  United States  2/19/1980   177.8      71   
3         Woodbridge    NJ    07095  United States  7/26/1951   220.9      70   
4      Daytona Beach    FL    32114  United States  8/31/1931   183.9      70   

    bmi                         email         phone  
0  19.6      ZoeWellish@superrito.com  951-719-9170  
1  19.2          PamelaSHill@cuvox.de  217-569-3204  
2  24.8          JaeMDebord@gustr.com  402-363-6804  
3  31.7     PhanBaLiem@jourrapide.com  732-636-8246  
4  26.4  RafaelCardosoCosta@gustr.com  386-334-5237

In [44]:
treatment.head()

fname       lname  hba1c_start  hba1c_end  hba1c_change treatment  \
0  Veronika    Jindrová         7.63       7.20          0.43   auralin   
1      Skye  Gormanston         7.97       7.62          0.35   auralin   
2    Sophia      Haugen         7.65       7.27          0.38   auralin   
3     Eddie      Archer         7.89       7.55          0.34   auralin   
4      Asia     Woźniak         7.76       7.37          0.39   auralin   

   start  end adverse_reaction  
0     41   48              NaN  
1     33   36              NaN  
2     37   42              NaN  
3     31   38              NaN  
4     30   36              NaN


## What is Assessing Data?
The data analysis process consists of the following steps

Asking Questions
Data Wrangling

### Gathering Data
     i.   CSV files
     ii.  APIs
     iii. Web Scraping
     iv.  Databases

### Assessing Data

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

### Types of Unclean Data

There are 2 kinds of unclean data

1. <u>Dirty Data</u> (Data with Quality issues):<br>
    Dirty data, also known as low quality data. Low quality data has content issues.<br>
2. <u>Messy Data</u> (Data with tidiness issues)<br>
    Messy data, also known as untidy data. Untidy data has structural issues.

### 1. Dirty Data

When the data has following kinds of issues:

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

This kind of data is known as Dirty data

### 2. Messy Data

When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

If a data violates any one of the above it is known as untidy data

Example:

<img src="untidy.png">

### Types of Assessment

There are 2 types of assessment styles

1. Manual - Looking through the data manually in google sheets
2. Programmatic - By using pandas functions such as info(), describe() or sample()

### Steps in Assessment

There are 2 steps involved in Assessment

1. Discover
2. Document

### Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues